### PR TITLE
Run a single wandb instance per TPU run

### PR DESCRIPTION
**Pull request by LysandreJik**

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -336,13 +336,14 @@ def _setup_wandb(self):
...
                 (Optional): boolean - defaults to false, set to "true" to disable wandb entirely
         """
-        logger.info('Automatic Weights & Biases logging enabled, to disable set os.environ["WANDB_DISABLED"] = "true"')
-        wandb.init(project=os.getenv("WANDB_PROJECT", "huggingface"), config=vars(self.args))
-        # keep track of model topology and gradients
-        if os.getenv("WANDB_WATCH") != "false":
-            wandb.watch(
-                self.model, log=os.getenv("WANDB_WATCH", "gradients"), log_freq=max(100, self.args.logging_steps)
-            )
+        if (is_tpu_available() and self.is_world_master()) or not is_tpu_available():
```
Comment (LysandreJik) [inline src/transformers/trainer.py:339]: @julien-c haven't tried with DP/DDP so I don't know if it should just be `self.is_world_master()`.

Have you already tried wandb on such setups before?

**Comment by borisdayma** [comment]
Why is it specific to tpu only?
Would the same logic apply in all cases?

**Comment by julien-c** [comment]
Yes I'm guessing we only want the global master to log to wandb in DDP, like we do for Tensorboard.

Not sure why we hadn't done it like that before, @borisdayma – thoughts?

File: src/transformers/trainer.py (hunk 1 above)
Comment (sshleifer) [reply src/transformers/trainer.py:339]: this is also a problem outside of tpu:
https://github.com/wandb/client/issues/452#issuecomment-557779621

Comment (julien-c) [reply src/transformers/trainer.py:339]: @sshleifer see https://github.com/huggingface/transformers/pull/4851#issuecomment-641587040

Comment (julien-c) [reply src/transformers/trainer.py:339]: good link, thanks for linking

**Comment by borisdayma** [comment]
I did not consider DP/DDP at the time. Actually Tensorboard logging does not consider world master either (only for logging config parameters but not metrics).

I understand we should wrap the entire `wandb` and Tensorboard logics within a simple `if self.is_world_master`.

May I suggest the following:

* refactor logging through PR #4756 
* add an equivalent `TFTrainer.is_world_master`
* wrap relevant Tensorboard & wandb sections of `log_metrics` by checking `is_world_master`
* call `setup_wandb` only for world master (checked either within `Trainer` & `TFTrainer` or within `setup_wandb`)

Let me know if you want me to add those changes.

**Pull request merged**